File: CivitasJava/civitas/SorpresaPorJugador.java
package civitas;

import java.util.ArrayList;

public class SorpresaPorJugador extends Sorpresa {
    private int valor;
    
    SorpresaPorJugador(int val, String tex){
        super(tex);
        valor = val;
    }
    
    @Override
    void aplicarAJugador (int actual, ArrayList<Jugador> todos) throws Exception{
        if ( jugadorCorrecto(actual, todos) ){
            informe(actual, todos);
			
            // Cobra al resto de jugadores
            String str = "Pagas " + valor + " al jugador " + todos.get(actual).getNombre() + "\n";
            SorpresaPagarCobrar pagar = new SorpresaPagarCobrar(valor * -1, str);
            
            for (int i = 0; i < todos.size(); i++)
                if (i != actual)
                    pagar.aplicarAJugador(i, todos);
			
            // Recibe el dinero
            str = "Recibes " + valor + " de cada jugador (" + valor*(todos.size()-1) + " en total)\n";
            SorpresaPagarCobrar cobrar = new SorpresaPagarCobrar(valor * (todos.size()-1), str);
            cobrar.aplicarAJugador(actual, todos);
        }
    }
}

File: CivitasRuby/sorpresa_por_jugador.rb
require_relative 'sorpresa'

module Civitas
class SorpresaPorJugador < Sorpresa
  
  def initialize(valor, texto)
    super(texto)
    @valor = valor
  end
  
  def aplicar_a_jugador(actual, todos)
    if jugador_correcto(actual, todos)
      informe(actual, todos)
      str = "Pagas " + @valor + " al jugador " + todos[actual].nombre
      sorpresa_pagar = Sorpresa.new_resto(TipoSorpresa::PAGARCOBRAR, @valor * -1, str)
      
      for i in 0..3
        if i != actual
          sorpresa_pagar.aplicar_a_jugador(i, todos)
        end
      end
      
      str = "Recibes " + @valor + " de cada jugador (" + @valor*(todos.size()-1) + " en total)"
      sorpresa_cobrar = Sorpresa.new_resto(TipoSorpresa::PAGARCOBRAR, @valor, str)
      sorpresa_cobrar.aaplicar_a_jugador(actual, todos)
    end
  end
  
  public_class_method :new
end
end
